test/boost/cql_query_test.cc

```diff
@@ -1091,6 +1092,32 @@ SEASTAR_TEST_CASE(test_range_deletion_scenarios_with_compact_storage) {
     });
 }
 
+SEASTAR_TEST_CASE(test_invalid_range_deletion) {
+    cql_test_config cfg;
+    cfg.disabled_features.insert(sstring(gms::features::UNBOUNDED_RANGE_TOMBSTONES));
+    return do_with_cql_env_thread([] (cql_test_env& e) {
+        cquery_nofail(e, "create table cf (p int, c1 int, c2 int, c3 int, primary key (p, c1, c2, c3));");
+        const auto q = [&] (const char* stmt) { return e.execute_cql(stmt).get(); };
+        using ire = exceptions::invalid_request_exception;
+        const auto expected = exception_predicate::message_contains("specify both bounds");
+        cquery_nofail(e, "delete from cf where p = 1");
+        cquery_nofail(e, "delete from cf where p = 1 and c1 = 2");
+        BOOST_REQUIRE_EXCEPTION(q("delete from cf where p = 1 and c1 < 2"), ire, expected);
+        BOOST_REQUIRE_EXCEPTION(q("delete from cf where p = 1 and c1 >= 0"), ire, expected);
+        cquery_nofail(e, "delete from cf where p = 1 and c1 < 2 and c1 >= 0");
+        // TODO: enable when supported:
+        // cquery_nofail(e, "delete from cf where p = 1 and c1 = 2 and c1 < 2");
```

Comment:
Do you really mean **c1** = 2 and **c1** < 2? Doesn't sound particularly useful, and I double-checked on Cassandra just in case - it's not possible to mix equality with other relations in order not to shoot yourself in the foot. Perhaps the idea was c1 = 2 and c2 < 2? Please clarify. Also, why should this particular line be supported, while the ones below won't?

Also, side note, `int`s of all types are perfectly bounded themselves, so we could just handle them without unbounded range tombstones if we wanted to. But that's not really that important and out of scope.

Reply:
> Do you really mean **c1** = 2 and **c1** < 2?

No, I meant what I wrote.  The error returned on this query is not about unbounded ranges but about mixing relations.  I think that should change in the future; no other programming or database language forbids multiple relations on the same variable.  Such queries should simply return their result (which may be empty, but so what -- many other queries can also return empty results and we still allow them to execute).  Cassandra's choice to forbid this is hard to understand, especially since Cassandra allows `c1>0 AND c1<0`, for example.  (See #5799.)

I don't think a human would ever deliberately craft this query, but most queries we serve are not crafted by humans; they are crafted by front-end applications.  If this query is invalid, that complicates such applications in a way that I suspect they don't want.

> Also, why should this particular line be supported, while the ones below won't?

By "supported" I mean "not rejected based on which names it has in its expressions".  All these queries should be supported in that sense.  If you're asking why this query doesn't expect the unbounded-range exception, it's because the range is empty. :)

Reply:
Alright